### PR TITLE
Better scalar multiplication for Short Weierstrass curves

### PR DESCRIPTION
**Pull request by Pratyush**
## Description

Uses affine oeprations, but in a way that is guaranteed to be safe. Tested against PR https://github.com/arkworks-rs/curves/pull/37, which tests scalar multiplication more comprehensively.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: src/groups/curves/short_weierstrass/non_zero_affine.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
+use super::*;
+/// An affine representation of a curve point that is guaranteed
```
Comment (ValarDragon) [inline src/groups/curves/short_weierstrass/non_zero_affine.rs:2]:
```suggestion
/// An affine representation of a prime order curve point that is guaranteed
```

File: src/groups/curves/short_weierstrass/non_zero_affine.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+                        .unwrap_or(P::BaseField::zero()))
+            })?;
+            lambda.mul_equals(&denominator, &numerator)?;
+            let x3 = lambda.square()? - x1 - x2;
+            let y3 = lambda * &(x1 - &x3) - y1;
+            Ok(Self::new(x3, y3))
+        }
+    }
+
+    /// Doubles `self`
```
Comment (ValarDragon) [inline src/groups/curves/short_weierstrass/non_zero_affine.rs:72]:
```suggestion
    /// Doubles `self`. As this a prime order curve point, this is guaranteed to not be the point at infinity.
```

File: src/groups/curves/twisted_edwards/mod.rs [hunk 3]
```diff
@@ -523,43 +523,35 @@ where
...
                 let x = F::two_bit_lookup(&bits, &x_s)?;
                 let y = F::two_bit_lookup(&bits, &y_s)?;
                 *self += Self::new(x, y);
-            } else if bits_base_powers.len() == 1 {
-                let bit = bits_base_powers[0].0.clone();
-                let base_power = bits_base_powers[0].1;
-                let new_encoded = &*self + base_power;
-                *self = bit.select(&new_encoded, &self)?;
+            } else if bits.len() == 1 {
+                // println!("Here");
```
Comment (ValarDragon) [inline src/groups/curves/twisted_edwards/mod.rs:551]: Debug code

File: src/groups/curves/short_weierstrass/non_zero_affine.rs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            let cs = [self, other].cs();
+            let (x1, y1) = (&self.x, &self.y);
+            let (x2, y2) = (&other.x, &other.y);
+            // Then,
+            // slope lambda := (y2 - y1)/(x2 - x1);
+            // x3 = lambda^2 - x1 - x2;
+            // y3 = lambda * (x1 - x3) - y1
+            let numerator = y2 - y1;
+            let denominator = x2 - x1;
+            let lambda = F::new_witness(ns!(cs, "lambda"), || {
```
Comment (ValarDragon) [inline src/groups/curves/short_weierstrass/non_zero_affine.rs:58]: `FpVar::mul_by_inverse` exists for this situation! Or is this being done for the `cs!` here?

Comment (ValarDragon) [reply src/groups/curves/short_weierstrass/non_zero_affine.rs:58]: Oh, someone changed mul by inverse to be two constraints. We should probably revert that change (or change the API to be `mul_by_inverse` and `mul_by_inverse_unchecked`. I guess we can do this in a separate PR then.

Comment (Pratyush) [reply src/groups/curves/short_weierstrass/non_zero_affine.rs:58]: `mul_by_inverse` for `Fpvar` still requires only a single constraint, right?

Comment (ValarDragon) [reply src/groups/curves/short_weierstrass/non_zero_affine.rs:58]: No, it was changed to 2 constraints: https://github.com/arkworks-rs/r1cs-std/blob/master/src/fields/fp/mod.rs#L730
(d.inverse() is a full constraint)

Comment (ValarDragon) [reply src/groups/curves/short_weierstrass/non_zero_affine.rs:58]: I'll make a PR to fix that

Comment (Pratyush) [reply src/groups/curves/short_weierstrass/non_zero_affine.rs:58]: Oh right. hmm

Comment (Pratyush) [reply src/groups/curves/short_weierstrass/non_zero_affine.rs]: Done, thanks for the heads up!

**Pull request merged**